### PR TITLE
[INV-20] Add guards and strategies for the access and refresh tokens

### PR DESCRIPTION
**Pull request by ShahobiddinIsmoilov**
Demonstration:

https://github.com/user-attachments/assets/915beb2f-4bd2-4e3b-ad00-5fae2adfc4b6

### CODE REVIEW

File: src/modules/auth/auth.service.ts [hunk 1]
```diff
@@ -104,37 +104,19 @@ export class AuthService {
...
-      throw new UnauthorizedException('Invalid or expired token');
-    }
-
-    const decodedRefreshToken =
-      this.jwtService.decode<AuthTokenPayload>(refreshToken);
+  async issueNewTokens(refreshToken: string, accountId: string) {
     const existingRefreshTokenEntity =
       await this.refreshTokenRepository.findOne({
         where: {
-          accountId: decodedRefreshToken.sub,
```
Comment (NegatiV-64) [inline src/modules/auth/auth.service.ts:121]: Please keep using sub. It's a common practice

File: src/modules/auth/auth.service.ts [hunk 2]
```diff
@@ -104,37 +104,19 @@ export class AuthService {
...
           token: refreshToken,
-          expiresAt: MoreThan(Date.now()),
           isDeleted: false,
         },
         relations: {
           account: true,
         },
       });
 
-    if (
```
Comment (NegatiV-64) [inline src/modules/auth/auth.service.ts:131]: Why it was removed?

File: src/modules/auth/strategies/access-token.strategy.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        infer: true,
+      }),
+    });
+  }
+
+  public async validate({ sub, userEmail, userRole }: AuthTokenPayload) {
+    const existingAccount = await this.accountRepository.findOneBy({ id: sub });
+
+    if (!existingAccount || existingAccount.isDeleted) {
+      throw new UnauthorizedException('Account was deleted or is inactive');
```
Comment (NegatiV-64) [inline src/modules/auth/strategies/access-token.strategy.ts:34]: Can we also check if user role is matching?

Comment (NegatiV-64) [reply src/modules/auth/strategies/access-token.strategy.ts:34]: And let's not say that user is deleted

File: src/shared/extractors/account-id.extractor.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import {
+  createParamDecorator,
+  ExecutionContext,
+  UnauthorizedException,
+} from '@nestjs/common';
+
+export const AccountId = createParamDecorator(
```
Comment (NegatiV-64) [inline src/shared/extractors/account-id.extractor.ts:7]: We have GetUser decorator, please update it instead

**Review by NegatiV-64 — CHANGES_REQUESTED** [review]

File: src/modules/auth/auth.controller.ts [hunk 5]
```diff
@@ -68,9 +69,14 @@ export class AuthController {
     status: 401,
     description: 'Invalid or expired token',
   })
-  async issueNewTokens(@Body() refreshTokenDto: RefreshTokenDto) {
+  @UseGuards(RefreshTokenGuard)
+  async issueNewTokens(
+    @Body() refreshTokenDto: RefreshTokenDto,
+    @GetUser('id') accountId: string,
```
Comment (NegatiV-64) [inline src/modules/auth/auth.controller.ts:75]: Maybe rename it to Get Account? WDYT?

File: src/modules/auth/decorators/get-user.decorator.ts [hunk 6]
```diff
@@ -5,11 +5,13 @@ import {
 } from '@nestjs/common';
 
 export const GetUser = createParamDecorator(
-  (data: unknown, ctx: ExecutionContext) => {
+  (data: string | undefined, ctx: ExecutionContext) => {
```
Comment (NegatiV-64) [inline src/modules/auth/decorators/get-user.decorator.ts:8]: Let's type it, what fields can be there

File: src/modules/auth/decorators/get-user.decorator.ts [hunk 7]
```diff
@@ -5,11 +5,13 @@ import {
...
     const request = ctx.switchToHttp().getRequest();
+
     if (!request.user) {
-      throw new UnauthorizedException('User not found');
+      throw new UnauthorizedException('User is not authenticated');
     }
-    return request.user;
+
+    return data ? request.user['data'] : request.user;
```
Comment (NegatiV-64) [inline src/modules/auth/decorators/get-user.decorator.ts:15]: What should happen if by data property there's nothing?

File: src/modules/auth/strategies/access-token.strategy.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+      }),
+    });
+  }
+
+  public async validate({ sub, userEmail, userRole }: AuthTokenPayload) {
+    const existingAccount = await this.accountRepository.findOneBy({ id: sub });
+
+    if (
+      !existingAccount ||
+      existingAccount.isDeleted ||
```
Comment (NegatiV-64) [inline src/modules/auth/strategies/access-token.strategy.ts:35]: You can simply remove deleted tokens using SQL

File: src/shared/extractors/request-body.extractor.ts
Comment (NegatiV-64) [inline src/shared/extractors/request-body.extractor.ts:1]: Is it used anywhere?

**Review by NegatiV-64 — CHANGES_REQUESTED** [review]

File: src/modules/auth/decorators/get-user.decorator.ts [hunk 9]
```diff
@@ -5,11 +5,13 @@ import {
 } from '@nestjs/common';
 
 export const GetUser = createParamDecorator(
-  (data: unknown, ctx: ExecutionContext) => {
+  (data: string | undefined, ctx: ExecutionContext) => {
     const request = ctx.switchToHttp().getRequest();
+
     if (!request.user) {
-      throw new UnauthorizedException('User not found');
+      throw new UnauthorizedException('User is not authenticated');
```
Comment (farrooh) [inline src/modules/auth/decorators/get-user.decorator.ts:12]: Please create error messages to use globally. There are the same type of exceptions that are using different messages right now. 
image

File: src/shared/extractors/request-body.extractor.ts
Comment (farrooh) [reply src/shared/extractors/request-body.extractor.ts:1]: Yeah, in RefreshTokenStrategy:
image

File: src/modules/auth/decorators/get-user.decorator.ts (hunk 6 above)
Comment (farrooh) [inline src/modules/auth/decorators/get-user.decorator.ts:8]: Specify keys. You can use `keyof` with a type for `req.user` if there is any.

**Review by farrooh — CHANGES_REQUESTED** [review]

File: src/modules/auth/decorators/get-user.decorator.ts (hunk 6 above)
Comment (farrooh) [reply src/modules/auth/decorators/get-user.decorator.ts:8]: You can use `keyof` with a type for `request.user` if there is any.

File: src/modules/auth/decorators/get-account.decorator.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+export const GetAccount = createParamDecorator(
+  (data: keyof RequestAccount | undefined, ctx: ExecutionContext) => {
+    const request = ctx.switchToHttp().getRequest();
+    const account = extractRequestAccount(request);
+
+    if (!account) {
+      throw new UnauthorizedException();
+    }
+
+    return data ? account[data] : account;
```
Comment (NegatiV-64) [inline src/modules/auth/decorators/get-account.decorator.ts:18]: What should happen if account[data] is undefined? Let's throw an error here

**Pull request merged**